### PR TITLE
Develop

### LINKED ISSUE #24: add thermal limit (or max rate) for all power lines.
When building a line or a transformer we need to make sure the rate is defined and below the SIL value. 

I am removing the ```Union{Nothing, Float64}``` in the rate field at branch types. 

see [cd32958](https://github.com/NREL/PowerSystems.jl/commit/cd329580a300d8686e28d63364585d4d086d0c22)

### LINKED ISSUE #34: dict_to_struct complains in Float -> Bool conversion
The `dict_to_struct` function is causing some errors where available is being passed as a Float64 and not a Bool. I added some convert in the code to fix it but requires review. Check  Develop. 

```Julia 
Error During Test
  Test threw an exception of type LoadError
  Expression: include("readforecastdata.jl")
  LoadError: MethodError: no method matching PowerSystems.RenewableCurtailment(::String, ::Int64, ::PowerSystems.Bus, ::Float64, ::PowerSystems.EconRenewable, ::TimeSeries.TimeArray{Union{Float64, Missings.Missing},1,DateTime,Array{Union{Float64, Missings.Missing},1}})
  Closest candidates are:
    PowerSystems.RenewableCurtailment(::String, ::Bool, ::PowerSystems.Bus, ::Float64, ::Union{PowerSystems.EconRenewable, Void}, ::TimeSeries.TimeArray) at /Users/jdlara/.julia/v0.6/PowerSystems/src/models/generation/renewable_generation.jl:46
  Stacktrace:
   [1] gen_dict_parser(::Dict{String,Any}) at /Users/jdlara/.julia/v0.6/PowerSystems/src/parsers/dict_to_struct.jl:228
   [2] ps_dict2ps_struct(::Dict{String,Any}) at /Users/jdlara/.julia/v0.6/PowerSystems/src/parsers/dict_to_struct.jl:17
   [3] include_from_node1(::String) at ./loading.jl:576
   [4] include(::String) at ./sysimg.jl:14
   [5] include_from_node1(::String) at ./loading.jl:576
   [6] include(::String) at ./sysimg.jl:14
   [7] process_options(::Base.JLOptions) at ./client.jl:305
   [8] _start() at ./client.jl:371
  while loading /Users/jdlara/.julia/v0.6/PowerSystems/test/readforecastdata.jl, in expression starting on line 10
ERROR: LoadError: There was an error during testing
while loading /Users/jdlara/.julia/v0.6/PowerSystems/test/runtests.jl, in expression starting on line 237
```

### PR DESCRIPTION
**Pull request by jd-lara**
Updates the testing of network elements and corrects errors in branch components modeling. 

Fixes #24
Fixes #34

### CODE REVIEW

File: src/models/branches/lines.jl [hunk 1]
```diff
@@ -1,13 +1,41 @@
+function CheckThermalLimits(name::String,connectionpoints::@NT(from::Bus,to::Bus), r::Float64, x::Float64,rate::@NT(from_to::Float64, to_from::Float64),anglelimits::@NT(max::Float64, min::Float64))
```
Comment (kdheepak) [inline src/models/branches/lines.jl]: Is this actually doing any checking? We should rename to `calculate_thermal_limits`.

File: src/models/branches/lines.jl [hunk 2]
```diff
@@ -1,13 +1,41 @@
...
     b::@NT(from::Float64, to::Float64) #[pu]
-    # TODO: add a rate and angle consistency check
-    rate::Float64 #[MVA]
+    # TODO: angle consistency check
+    rate::@NT(from_to::Float64, to_from::Float64)
     anglelimits::@NT(max::Float64, min::Float64)
+
+    function Line(name, available, connectionpoints, r, x, b, rate, anglelimits)
+
+        rating =  CheckThermalLimits(name, connectionpoints, r, x,  @NT(from_to = rate, to_from = rate), anglelimits)
```
Comment (kdheepak) [inline src/models/branches/lines.jl]: Rename to `calculate_thermal_limits`.

Comment (jdlara-berkeley) [reply src/models/branches/lines.jl]: @sourabhdalvi move this to the checks in utils and rename please

File: src/models/branches/lines.jl [hunk 3]
```diff
@@ -1,13 +1,41 @@
...
     name::String
     available::Bool
     connectionpoints::@NT(from::Bus, to::Bus)
     r::Float64 #[pu]
     x::Float64 #[pu]Co
     b::@NT(from::Float64, to::Float64) #[pu]
-    # TODO: add a rate and angle consistency check
-    rate::Float64 #[MVA]
+    # TODO: angle consistency check
+    rate::@NT(from_to::Float64, to_from::Float64)
```
Comment (claytonpbarrows) [inline src/models/branches/lines.jl]: can we put the unit comment back?

Comment (jd-lara) [reply src/models/branches/lines.jl]: Done

File: src/models/branches/lines.jl (hunk 1 above)
Comment (jd-lara) [reply src/models/branches/lines.jl]: This has been addressed with the latest commits 
bbf9a0f

File: src/utils/system_checks.jl [hunk 4]
```diff
@@ -59,13 +61,48 @@ function PVBusCheck(buses::Array{Bus}, generators::Array{T}) where {T<:Generator
...
         if b.bustype == "PV"
-            b.number in pv_list ? continue : error("The bus ", b.number, " is declared as PV without a generator connected to it")
+            b.number in pv_list ? continue : warn("The bus ", b.number, " is declared as PV without a generator connected to it")
         else
             continue
         end
     end
 end
 
+function check_angle_limits!(anglelimits::@NT(max::Float64, min::Float64))
```
Comment (kdheepak) [inline src/utils/system_checks.jl]: This function should either 

1) Take a namedtuple `anglelimits` as an argument and check if it is within limits, i.e. return true or false, and not mutate the argument (not recommended).
2) Take a namedtuple `anglelimits` as an argument and create a new tuple `anglelimits` which will allows have allowable limits, and return the new `anglelimits` (preferred).

NamedTuples to my understanding are immutable, so the `!` suffix is not recommended.

File: src/utils/system_checks.jl [hunk 5]
```diff
@@ -59,13 +61,48 @@ function PVBusCheck(buses::Array{Bus}, generators::Array{T}) where {T<:Generator
...
+    (anglelimits.max >= 90.0 && anglelimits.min <= -90.0) ? anglelimits = @NT(max = 90.0, min = -90.0) : true
+    (anglelimits.max >= 90.0 && anglelimits.min >= -90.0) ? anglelimits = @NT(max = 90.0, min = anglelimits.min) : true
+    (anglelimits.max <= 90.0 && anglelimits.min <= -90.0) ? anglelimits = @NT(max = anglelimits.max, min = -90.0) : true
+    (anglelimits.max == 0.0 && anglelimits.min == 0.0) ? anglelimits = @NT(max = 90.0, min = -90.0): true
+
+    return anglelimits
+
+end
+
+function calculate_thermal_limits!(r::Float64, x::Float64, rate::@NT(from_to::Float64, to_from::Float64), anglelimits::@NT(max::Float64, min::Float64), connectionpoints::@NT(from::Bus, to::Bus))
```
Comment (kdheepak) [inline src/utils/system_checks.jl]: Remove `!`. This function is not mutating any argument.

**Comment by jd-lara** [comment]
There is an issue with the data.zip file, @sourabhdalvi can you take a look? Otherwise the PR is passing tests in my computer

**Comment by kdheepak** [comment]
It looks good! Once the tests pass this can be merged.

**Comment by jd-lara** [comment]
The tests are failing when downloading the data.zip.

**Pull request merged**